### PR TITLE
Music: post milestones without page reloads

### PR DESCRIPTION
**Pull request by breville**
With this change, passing a Music Lab level triggers a "milestone" post to the server, which marks the level as complete, and turns the bubble in the header to green.  This is a follow-up to https://github.com/code-dot-org/code-dot-org/pull/50923, which added the ability to switch music levels in a script without page reloads.  

It's worth noting that opening the `HeaderPopup` to reveal the full script's lessons triggers a `/api/user_progress` network request, which asynchronously returns the user's progress from the server and is then merged in (even if the popup has been closed by the user before the request completes).  In https://github.com/code-dot-org/code-dot-org/pull/50923, this would lead to some issues in which the active level might change unexpectedly, and the various assumptions that we had a single lesson's worth of progress were no longer correct.  These are now addressed in this change: we now use the `currentLevelId` in the progress store if it's available, and we now search for the appropriate lesson in the local store.

https://user-images.githubusercontent.com/2205926/234437793-7f8adbb8-92fa-454c-9258-2c7055c36402.mov

### CODE REVIEW

File: apps/src/music/progress/ProgressManager.ts [hunk 1]
```diff
@@ -103,9 +103,11 @@ export default class ProgressManager {
         // Ask the lab-specific validator if this validation's
         // conditions are met.
         if (this.validator.conditionsMet(validation.conditions)) {
-          this.currentProgressState.satisfied = validation.next;
-          this.currentProgressState.message = validation.message;
-          this.onProgressChange();
+          if (!this.currentProgressState.satisfied) {
```
Comment (molly-moen) [inline apps/src/music/progress/ProgressManager.ts:106]: what is the reason for this change?

Comment (breville) [reply apps/src/music/progress/ProgressManager.ts:106]: Oh, we were calling `this.onProgressChange()` repeatedly once the criteria were satisfied.  This change ensures we only call it once, the first time they are satisfied.

File: apps/src/code-studio/progressRedux.js [hunk 2]
```diff
@@ -84,18 +85,51 @@ const initialState = {
...
+// so we should update the browser and also set this as the new
+// current level.
 export function navigateToLevelId(levelId) {
   return (dispatch, getState) => {
     const state = getState().progress;
-    const newLevel = state.lessons[0].levels.find(level =>
-      level.ids.find(id => id === levelId)
-    );
+    const newLevel = state.lessons
+      .find(lesson => lesson.id === state.currentLessonId)
```
Comment (sanchitmalhotra126) [inline apps/src/code-studio/progressRedux.js]: minor: handle the case where `.find()` returns null?

File: apps/src/code-studio/progressRedux.js [hunk 3]
```diff
@@ -84,18 +85,51 @@ const initialState = {
...
 
+// The user has successfully completed the level and the page
+// will not be reloading.
+export function sendSuccessReport(appType) {
+  return (dispatch, getState) => {
+    const state = getState().progress;
+    const levelId = state.currentLevelId;
+    const currentLevel = state.lessons
+      .find(lesson => lesson.id === state.currentLessonId)
+      .levels.find(level => level.ids.find(id => id === levelId));
```
Comment (sanchitmalhotra126) [inline apps/src/code-studio/progressRedux.js]: nit: pull this out to common helper function?

File: apps/src/code-studio/progressRedux.js [hunk 4]
```diff
@@ -84,18 +85,51 @@ const initialState = {
...
+    const scriptLevelId = currentLevel.id;
+
+    const data = {
+      app: appType,
+      result: true,
+      testResult: 100,
```
Comment (sanchitmalhotra126) [inline apps/src/code-studio/progressRedux.js]: For my understanding, anything special about this number? Is it always 100 whenever a milestone is completed?

File: apps/src/code-studio/progressRedux.js [hunk 5]
```diff
@@ -84,18 +85,51 @@ const initialState = {
...
+    };
+
+    return $.ajax({
```
Comment (sanchitmalhotra126) [inline apps/src/code-studio/progressRedux.js]: can we use `fetch` instead of jquery?

File: apps/src/code-studio/progressRedux.js [hunk 6]
```diff
@@ -84,18 +85,51 @@ const initialState = {
...
+      type: 'POST',
+      url: `/milestone/0/${scriptLevelId}/${levelId}`,
```
Comment (sanchitmalhotra126) [inline apps/src/code-studio/progressRedux.js]: What's the `0` here?

File: apps/src/code-studio/progressRedux.js [hunk 7]
```diff
@@ -109,7 +143,7 @@ export default function reducer(state = initialState, action) {
     // extract fields we care about from action
     return {
       ...state,
-      currentLevelId: action.currentLevelId,
+      currentLevelId: state.currentLevelId || action.currentLevelId,
```
Comment (sanchitmalhotra126) [inline apps/src/code-studio/progressRedux.js:161]: Any chance this might have unintended side effects? how often do we dispatch the INIT_PROGRESS action?

**Review by sanchitmalhotra126 — COMMENTED** [review]
Few questions, nothing major!

File: apps/src/code-studio/progressRedux.js (hunk 4 above)
Comment (breville) [reply apps/src/code-studio/progressRedux.js]: It's an "ideal pass".  I'll update with the official constant from [here](https://github.com/code-dot-org/code-dot-org/blob/49438f119d0551b5aa4c23b0f4860b395696a463/apps/src/constants.js#L100), which corresponds to the server-side constant [here](https://github.com/code-dot-org/code-dot-org/blob/49438f119d0551b5aa4c23b0f4860b395696a463/lib/cdo/activity_constants.rb#L17).

File: apps/src/code-studio/progressRedux.js (hunk 6 above)
Comment (breville) [reply apps/src/code-studio/progressRedux.js]: In the [route](https://github.com/code-dot-org/code-dot-org/blob/49438f119d0551b5aa4c23b0f4860b395696a463/dashboard/config/routes.rb#L524) it's the user ID, though in other labs we also seem to let the server determine this and simply pass 0.  I'll make this a constant with a comment.

File: apps/src/code-studio/progressRedux.js (hunk 7 above)
Comment (breville) [reply apps/src/code-studio/progressRedux.js:161]: It's a healthy concern.  It's dispatched once when the header is first built at page load, using static data baked into the page source, and again each time the user opens the `HeaderPopup` menu to see all levels (at the completion of the asynchronous `/api/user_progress` network request that is triggered by opening that popup).  We want to use `action.currentLevelId` that first time, since it's all we have and is correct at the time, and then `state.currentLevelId` for the subsequent times, just in case the user has switched level without a page reload.

File: apps/src/code-studio/progressRedux.js (hunk 5 above)
Comment (breville) [reply apps/src/code-studio/progressRedux.js]: Will do!

File: apps/src/code-studio/progressRedux.js (hunk 3 above)
Comment (breville) [reply apps/src/code-studio/progressRedux.js]: Will do!

File: apps/src/code-studio/progressRedux.js (hunk 2 above)
Comment (breville) [reply apps/src/code-studio/progressRedux.js]: Agreed that this should be handled somehow, but also kind of torn, because if we handle it silently on the client only, we essentially mask an inability to report a milestone.  If we were to report it, we'd use one of the available reporting mechanisms; you probably have some ideas given your other current work?

Given that we deploy so often, I'd think that New Relic alerting us to the JS error would actually be a very effective way to catch an issue here, except that we have the issue with so many errors in there that we don't get very effective alerting from it...

Comment (sanchitmalhotra126) [reply apps/src/code-studio/progressRedux.js]: Yeah, if we want to flag this as invalid state from the server somehow, then logging would be best. If we want to take advantage of the new browser logging, maybe we can just create a TODO and link a task here?

File: apps/src/code-studio/progressRedux.js [hunk 8]
```diff
@@ -40,8 +40,9 @@ const OVERWRITE_RESULTS = 'progress/OVERWRITE_RESULTS';
 const PEER_REVIEW_ID = -1;
 
 const initialState = {
-  // These first fields never change after initialization
   currentLevelId: null,
+
+  // These first fields never change after initialization
```
Comment (sanchitmalhotra126) [inline apps/src/code-studio/progressRedux.js:45]: This is an interesting comment 🙂 seems like a good reason to pull these values out of this store at some point...

File: apps/src/code-studio/progressRedux.js [hunk 9]
```diff
@@ -84,18 +85,65 @@ const initialState = {
...
+    const state = getState().progress;
+    const levelId = state.currentLevelId;
+    const currentLevel = getLevelById(
+      state.lessons,
+      state.currentLessonId,
+      levelId
+    );
+    const scriptLevelId = currentLevel.id;
+
+    // The server can determine the user ID.
```
Comment (sanchitmalhotra126) [inline apps/src/code-studio/progressRedux.js]: nit: can we something like 'a value of 0 indicates to the server to load the user ID from the current session' just to be clear?

File: apps/src/code-studio/progressRedux.js [hunk 10]
```diff
@@ -84,18 +85,65 @@ const initialState = {
...
+
+    const data = {
+      app: appType,
+      result: true,
+      testResult: idealPassResult,
+    };
+
+    fetch(`/milestone/${userId}/${scriptLevelId}/${levelId}`, {
+      method: 'POST',
+      headers: {
```
Comment (sanchitmalhotra126) [inline apps/src/code-studio/progressRedux.js:134]: Interesting given the discussion we were having around CSRF that this endpoint has no CSRF token requirement...seems like this is [intentionally disabled](https://github.com/code-dot-org/code-dot-org/blob/cfa114efdb6cd0d41ebeba9895e7226b93d4752a/dashboard/app/controllers/activities_controller.rb#L12) in the controller, but i wonder if we should now use the ajax method of fetching the token here? Not at all required for this change, but just thinking out loud.

File: apps/src/code-studio/progressRedux.js (hunk 7 above)
Comment (sanchitmalhotra126) [reply apps/src/code-studio/progressRedux.js:161]: Gotcha, thanks! This seems like a potential use case for GraphQL, maybe we would only request certain parts of the progress model when opening the header popup

**Review by sanchitmalhotra126 — APPROVED** [review]
LGTM!

**Review by molly-moen — APPROVED** [review]

File: apps/src/code-studio/progressRedux.js (hunk 9 above)
Comment (breville) [reply apps/src/code-studio/progressRedux.js]: From perusing the [controller](https://github.com/code-dot-org/code-dot-org/blob/cfa114efdb6cd0d41ebeba9895e7226b93d4752a/dashboard/app/controllers/activities_controller.rb#L22), it appears that the server doesn't even use this parameter anymore...

Comment (breville) [reply apps/src/code-studio/progressRedux.js]: I've added a comment to that effect.

File: apps/src/code-studio/progressRedux.js (hunk 2 above)
Comment (breville) [reply apps/src/code-studio/progressRedux.js]: Opened a tracking issue [here](https://codedotorg.atlassian.net/browse/SL-823).
